pkg/apis/storage/validation/validation.go

```diff
@@ -266,3 +279,176 @@ func validateAllowedTopologies(topologies []api.TopologySelectorTerm, fldPath *f
+			idxPath := fldPath.Index(i)
+			allErrs = append(allErrs, field.Duplicate(idxPath.Child("name"), driver.Name))
+		}
+		driverNamesInStatuses.Insert(driver.Name)
+	}
+
+	return allErrs
+}
+
+// validateCSIDriverInfoSpecNodeID tests if Name in CSIDriverInfoSpec is a valid node id.
+func validateCSIDriverInfoSpecNodeID(nodeID string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(nodeID) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, nodeID))
+	}
+	return allErrs
+}
+
+// validateCSIDriverInfoSpecTopologyKeys tests if TopologyKeys in CSIDriverInfoSpec are
+// valid topology keys.
+func validateCSIDriverInfoSpecTopologyKeys(keys []string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for _, key := range keys {
+		if len(key) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath, key))
+		}
+	}
+	return allErrs
+}
+
+// validateCSIDriverInfoStatusVolumePluginMechanism tests if VolumePluginMechanism in
+// CSIDriverInfoStatus is valid
+func validateCSIDriverInfoStatusVolumePluginMechanism(volumePluginMechanism string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if !supportedVolumePluginMechanisms.Has(volumePluginMechanism) {
+		allErrs = append(allErrs, field.NotSupported(fldPath, volumePluginMechanism, supportedVolumePluginMechanisms.List()))
+	}
+
+	return allErrs
+}
+
+// ValidateCSIDriver validates a CSIDriver.
+func ValidateCSIDriver(csiDriver *storage.CSIDriver) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, apivalidation.ValidateCSIDriverName(csiDriver.Name, field.NewPath("name"))...)
+
+	allErrs = append(allErrs, validateCSIDriverSpec(&csiDriver.Spec, field.NewPath("spec"))...)
+	return allErrs
+}
+
+// ValidateCSIDriverUpdate validates a CSIDriver.
+func ValidateCSIDriverUpdate(new, old *storage.CSIDriver) field.ErrorList {
+	allErrs := ValidateCSIDriver(new)
+
+	// Spec is read-only
+	// If this ever relaxes in the future, make sure to increment the Generation number in PrepareForUpdate
+	if !apiequality.Semantic.DeepEqual(old.Spec, new.Spec) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), new.Spec, "field is immutable"))
+	}
+	return allErrs
+}
+
+// ValidateCSIDriverSpec tests that the specified CSIDriverSpec
+// has valid data.
+func validateCSIDriverSpec(
+	spec *storage.CSIDriverSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, validatePodInfoOnMountVersion(*spec.PodInfoOnMountVersion, fldPath.Child("podInfoOnMountVersion"))...)
+	return allErrs
+}
+
+// validatePodInfoOnMountVersion tests if the podInfoOnMountVersion is valid for CSIDriver.
+func validatePodInfoOnMountVersion(podInfoOnMountVersion string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(podInfoOnMountVersion) > 0 && !supportedPodInfoOnMountVersions.Has(string(podInfoOnMountVersion)) {
```

Comment:
Now that it's a pointer, do we want to allow an empty string?

Reply:
The only valid value is "V1". So I think an empty string should be invalid, same as any other string that is not "V1".